### PR TITLE
feat(character): Add option for different symbol when root user [Draft]

### LINKED ISSUE #2421: Ability to set different character module symbol for normal and root user
#### Describe the solution you'd like
Would be nice to have this feature to reflect more clear difference of prompt that you're in. For example when user is root then set `#` symbol as character and when is not then something else. I don't think that there are needed any default values I guess it makes sense for this feature to be disabled by default and override current behavior by user's explicit choice.

### PR DESCRIPTION
**Pull request by DavidSmith166**
#### Description

Adds ability to set different character module symbol for normal and root user as described in #2421 

I'm still very much learning Rust so I don't know if my approach follows Starship's design philosophy. Any and all feedback would be greatly appreciated.

- Added root_symbol option in character config
- Added fn is_root_euid which determines whether the user's euid is 0 (root)
- Edited existing tests such that their behavior does not change with my changes

#### Motivation and Context

Closes #2421 

#### Screenshots (if appropriate):
starship_demo

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

File: src/modules/character.rs [hunk 1]
```diff
@@ -69,6 +82,25 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
...
+            }
+        },
+        None => None,
+    };
+
+    match euid {
+        Some(euid) => euid == root_uid,
+        None => false,
+    }
+}
```
Comment (davidkna) [inline src/modules/character.rs:102]: You should reuse `is_root_user()` from the `username` module here (`src/modules/username.rs`).

File: src/modules/character.rs [hunk 2]
```diff
@@ -113,6 +145,7 @@ mod test {
                 .config(toml::toml! {
                     [character]
                     success_symbol = "[➜](bold green)"
+                    root_symbol = "[➜](bold green)"
```
Comment (daniel-white) [inline src/modules/character.rs:148]: can we have `root_success_symbol` and `root_error_symbol`?

Comment (daniel-white) [reply src/modules/character.rs:148]: also for convention sake - can we change this to `#`

File: src/modules/character.rs (hunk 1 above)
Comment (DavidSmith166) [reply src/modules/character.rs]: Thank you for your feedback! I've implemented your suggestion

File: src/modules/character.rs (hunk 2 above)
Comment (DavidSmith166) [reply src/modules/character.rs]: Good call, how does it look now?

**Comment by daniel-white** [comment]
Is this restricted to Linux only?

File: src/configs/character.rs [hunk 3]
```diff
@@ -19,6 +22,9 @@ impl<'a> Default for CharacterConfig<'a> {
             success_symbol: "[❯](bold green)",
             error_symbol: "[❯](bold red)",
             vicmd_symbol: "[❮](bold green)",
+            root_success_symbol: "[#](bold green)",
+            root_error_symbol: "[#](bold red)",
+            root_symbol_enabled: false,
```
Comment (matchai) [inline src/configs/character.rs:27]: I'm a little concerned by the amount of added configuration options something like this adds.

Ideally, we would have `success_style`, `error_style`, `symbol` and `root_symbol`; similarly to how we have `style`, `not_capable_style`, and `symbol` for the Node.js module. The trade-off there being that folks wouldn't be able to change the error symbol. 🤔

I'd be curious to hear the thoughts of other @starship/astronauts about how we could go about handling these cases.

File: src/modules/character.rs [hunk 4]
```diff
@@ -28,14 +32,19 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
     let keymap = props.get("keymap").unwrap_or(&keymap_default);
     let exit_success = exit_code == "0";
 
+    let assumed_mode = match config.root_symbol_enabled && cfg!(target_os = "linux") && is_root_user() {
```
Comment (daniel-white) [inline src/modules/character.rs:35]: does this imply linux only - i'd love to see this on mac

Comment (kidonng) [reply src/modules/character.rs:35]: macOS's root user is disabled by default, `sudo` gives your **current user** root privileges, so this doesn't make much sense on macs.

https://support.apple.com/guide/directory-utility/about-the-root-user-dirub32398f1/mac

Comment (daniel-white) [reply src/modules/character.rs:35]: but on macs if i `sudo bash` i get a prompt with a `#`

Comment (kidonng) [reply src/modules/character.rs:35]: Root privileges ≠root user

Comment (davidkna) [reply src/modules/character.rs:35]: Even if using the root user is discouraged in macOS, I don't think the prompt should artificially restrict displaying that. The current implementation also seems to detect root even if the user is only acting with root privileges (`sudo bash` ).

Furthermore this check  shouldn't exclude all the other BSDs. `is_root_user()` is already handling different operating systems.

**Comment by aw1cks** [comment]
I would love to see this implemented, is there anything I can do to help with this?

**Pull request closed without merging**